app/src/test/kotlin/io/homeassistant/companion/android/widgets/todo/TodoWidgetToggleActionsTest.kt

```diff
@@ -0,0 +1,124 @@
+        val widgetId = 1
+
+        every { action.getEntryPoints(any()) } returns entryPoints
+        every { action.getGlanceManager(any()) } returns glanceManager
+        every { glanceManager.getAppWidgetId(any()) } returns widgetId
+
+        coEvery { entryPoints.dao().get(widgetId) } returns null
+
+        action.onAction(mockk(), FakeGlanceId(widgetId), parameters)
+
+        verify(exactly = 0) {
+            parameters[TOGGLE_KEY]
+            entryPoints.serverManager() wasNot called
+        }
+        coVerify(exactly = 1) {
+            entryPoints.dao().get(widgetId)
+        }
+    }
+
+    @Test
+    fun `Given a widgetID and todo item state without uid when present in DAO and invoking onAction then do nothing`() = runTest {
+        val action = spyk<ToggleTodoAction>()
+        val glanceManager = mockk<GlanceAppWidgetManager>()
+        val widgetId = 1
+        val todoItem = mockk<TodoWidgetEntity>()
+        val parameters = actionParametersOf(TOGGLE_KEY to TodoItemState(null, "", false))
+
+        every { action.getEntryPoints(any()) } returns entryPoints
+        every { action.getGlanceManager(any()) } returns glanceManager
+        every { glanceManager.getAppWidgetId(any()) } returns widgetId
+
+        coEvery { entryPoints.dao().get(widgetId) } returns todoItem
+
+        action.onAction(mockk(), FakeGlanceId(widgetId), parameters)
+
+        verify {
+            entryPoints.serverManager() wasNot called
+        }
+        coVerify(exactly = 1) {
+            entryPoints.dao().get(widgetId)
+        }
+    }
+
+    @Test
+    fun `Given a widgetID and todo item state with uid when present in DAO and invoking onAction then do updateTodo and toggle status and invoke update`() = runTest {
+        val action = spyk<ToggleTodoAction>()
+        val glanceManager = mockk<GlanceAppWidgetManager>()
+        val widgetId = 1
+        val todoItem = TodoWidgetEntity(1, 42, "HA")
+        val parameters = actionParametersOf(TOGGLE_KEY to TodoItemState("42", "", false))
+
+        every { action.getEntryPoints(any()) } returns entryPoints
+        every { action.getGlanceManager(any()) } returns glanceManager
+        every { glanceManager.getAppWidgetId(any()) } returns widgetId
+        coEvery {
+            entryPoints.webSocketRepository.updateTodo(any(), any(), any(), any())
+        } returns true
+
+        coEvery { entryPoints.dao().get(widgetId) } returns todoItem
+
+        // This is useful to validate that it calls update()
+        assertThrows<IllegalArgumentException>("Invalid Glance ID") {
+            action.onAction(mockk(), FakeGlanceId(widgetId), parameters)
+        }
+
+        coVerify(exactly = 1) {
+            entryPoints.dao().get(widgetId)
+            entryPoints.serverManager().webSocketRepository(42)
+            entryPoints.webSocketRepository.updateTodo("HA", "42", null, COMPLETED_STATUS)
+        }
+    }
+
+    @Test
+    fun `toggleStatus mapping is valid`() {
```

Comment:
Looks like this line is missing the Given, When Then format :)

Reply:
Fixed